src/hardware/mixer.cpp

```diff
@@ -789,6 +794,32 @@ AudioFrame MixerChannel::ApplyCrossfeed(const AudioFrame &frame) const
 	return {a.left + b.left, a.right + b.right};
 }
 
+void MixerChannel::WakeUp()
+{
+	autosleep = {};
+
+	if (!is_enabled)
+		Enable(true);
+}
+
+void MixerChannel::AccumulateAutoSleepFrame(const AudioFrame &frame)
+{
+	autosleep.accumulator += static_cast<int16_t>(fabsf(frame.left)) +
+	                         static_cast<int16_t>(fabsf(frame.right));
```

Comment:
We're getting into micro-optimisation territory here, and ultimately I don't think either solution would make a difference at all, but wouldn't it be simpler to just keep the accumulator a float? If you want to count clock cycles, I think the float<->conversions take more time than just using the same throughout, no?

Reply:
It's actually a bit more work doing this instead of just accumulating floats, which I originally did.

In the earlier versions, adding the "silent" floats actually bring in values on the order of `0.002` per-channel, per frame (lol; I still my Python window open where I calculated these!). 

Over a ~200ms, span, the accumulator will have a value on the order of about ~49 for a 48 KHz channel.

The higher the frequency, the more "quiet" samples accumulate; so I actually had a "ConfigureAutoSleep()" function in the SetSamplerate() function that set a reasonable "accumulated silent threshold" value to compare against.

The problem is.. we have some audio devices that can "channel->AddSilence()", and for those, no processing happens, and for these, we'd need to differentiate an actual sample on the order of ~50 versus thousands of below-1.0f samples that merely accumulated to ~50.

Fortunately the casting approach (truncating below-1), and then doing a boolean "zero-non-zero" compare on the accumulator gives us a strong guarantee that the channel's been silent.

All it takes is a single sample just above 1.0f to fail that case and keep the channel alive.

Reply:
Okay, so if I'm understanding you correctly, there's always some small residual output generated by the emulation, even if there is no sound output. Might be some minor DC offset or something. In any case, if this works, then it's all good, and thanks for the explanation 😄

Reply:
😅 yes; the inbound floats, even when "silent", each have a tiny bit of inaccuracy (just by nature of being floats), which is perfectly fine when the samples are assessed individually (the tiny bit of float inaccuracy doesn't come close to pushing the sample by a whole digit); but adding the residual float junk from 50,000 of them does eventually add up.

https://randomascii.wordpress.com/2012/02/25/comparing-floating-point-numbers-2012-edition/

Reply:
Yeah, but it's not really inaccuracy; you can keep adding 0.0f to itself as long as you want and it will stay exactly 0.0f. All integers have 100% exact float representations, so this must be some residual output issue, which is a bit weird, but OK, the fix is easy.